source/lac/petsc_solver.cc

```diff
@@ -200,39 +203,44 @@ namespace PETScWrappers
 
 
   void
-  SolverBase::initialize(const PreconditionBase &preconditioner)
+  SolverBase::create_with_comm(const MPI_Comm &comm)
   {
-    PetscErrorCode ierr;
+    // Create the PETSc KSP object
+    AssertPETSc(KSPCreate(comm, &ksp));
 
-    solver_data = std::make_unique<SolverData>();
+    // then a convergence monitor
+    // function that simply
+    // checks with the solver_control
+    // object we have in this object for
+    // convergence
+    ensure_convergence_test();
+  }
 
-    ierr = KSPCreate(preconditioner.get_mpi_communicator(), &solver_data->ksp);
-    AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+
+  void
+  SolverBase::ensure_convergence_test() const
+  {
+    if (ksp && solver_control)
+      AssertPETSc(
+        KSPSetConvergenceTest(ksp, &convergence_test, solver_control, nullptr));
+  }
```

Comment:
I think this would be a better name:
```suggestion
  void
  SolverBase::perhaps_set_convergence_test() const
  {
    if (ksp && solver_control)
      AssertPETSc(
        KSPSetConvergenceTest(ksp, &convergence_test, solver_control, nullptr));
  }
```
with the documentation stating that whether or not the function sets a convergence test depends on whether the object has been created with a `SolverControl` object.